abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+Creates a new timer. When a timer is created as a one-time alarm (`one_time`),
+then callback will be called only once after `period` milliseconds. Otherwise,
+the callback is going to be called every `period` milliseconds until the timer
+is deleted using `proxy_delete_timer` with the returned unique timer identifier
+(`return_timer_id`).
+
+Returns `Ok` on success.
+
+Returns `InvalidMemoryAccess` when `return_timer_id` points to invalid memory
+address.
+
+#### `proxy_delete_timer`
+
+* params:
+  - `i32 (uint32_t) timer_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created timer (`timer_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the timer (`timer_id`) doesn't exist.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_timer_ready`
+
+* params:
+  - `i32 (uint32_t) timer_id`
+* returns:
+  - none
+
+Called after configured period by the timer (`timer_id`) created using
+`proxy_create_timer`.
+
+
+## Stats/metrics
+
+### Functions exposed by the host environment
+
+#### `proxy_create_metric`
+
+* params:
+  - `i32 (proxy_metric_type_t) metric_type`
+  - `i32 (const char*) metric_name_data`
+  - `i32 (size_t) metric_name_size`
+  - `i32 (uint32_t*) return_metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Creates metric (`metric_name_data`, `metic_name_size`) of type `metric_type`.
+
+It can be referred to in `proxy_get_metric_value`, `proxy_set_metric_value`,
+and `proxy_increment_metric_value` using returned unique metric identifier
+(`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `metric_type`.
+
+Returns `InvalidMemoryAccess` when `metric_name_data`, `metric_name_size`
+and/or `return_metric_id` point to invalid memory address.
+
+#### `proxy_get_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i32 (uint64_t*) return_value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get the value of the metric (`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+#### `proxy_set_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (uint64_t) value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set the value of the metric (`metric_id`) to the value (`value`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when the `metric_type` doesn't support setting
+values.
+
+#### `proxy_increment_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (int64_t) offset`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Increment/decrement value of the metric (`metric_id`) by offset (`offset`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when `metric_type` doesn't support incrementing
+and/or decrementing values.
+
+#### `proxy_delete_metric`
```

Comment:
In theory, it's here to remove metrics from the host and avoid resource leaks, but I'm not sure how useful this will be in practice. It probably should match number of `proxy_create_metric` calls, but I could be convinced otherwise.